tests/test_send_recv_two_workers.py

```diff
@@ -0,0 +1,274 @@
+                        frames.append(b"")
+
+        msg = await from_frames(frames)
+
+        close_msg = b"shutdown listener"
+        close_msg_size = np.array([len(close_msg)], dtype=np.uint64)
+
+        await ep.send(close_msg_size)
+        await ep.send(close_msg)
+
+        print("Shutting Down Client...")
+        return msg["data"]
+
+    rx_cuda_obj = asyncio.get_event_loop().run_until_complete(read())
+    # nvlink only measures in KBs
+    num_bytes = nbytes(rx_cuda_obj)
+    print(f"TOTAL DATA: {num_bytes}")
+    if num_bytes > 1000:
+        rx, tx = total_nvlink_transfer()
+        msg = f"RX BEFORE SEND: {before_rx} -- RX AFTER SEND: {rx} \
+               -- TOTAL DATA: {num_bytes}"
+        print(msg)
+        assert rx > before_rx
+
+    cuda_obj_generator = cloudpickle.loads(func)
+    pure_cuda_obj = cuda_obj_generator()
+
+    from cudf.tests.utils import assert_eq
+    import cupy
+
+    if hasattr(rx_cuda_obj, "shape"):
+        shape = rx_cuda_obj.shape
+    else:
+        # handle an sr._column object
+        shape = (1,)
+    if len(shape) == 1:
+        cupy.testing.assert_allclose(rx_cuda_obj, pure_cuda_obj)
+    else:
+        assert_eq(rx_cuda_obj, pure_cuda_obj)
+
+
+def server(env, port, func):
+    # create listener receiver
+    # write cudf object
+    # confirm message is sent correctly
+
+    os.environ.update(env)
+    create_cuda_context()
+
+    async def f(listener_port):
+        # coroutine shows up when the client asks
+        # to connect
+        async def write(ep):
+
+            print("CREATING CUDA OBJECT IN SERVER...")
+            cuda_obj_generator = cloudpickle.loads(func)
+            cuda_obj = cuda_obj_generator()
+            msg = {"data": to_serialize(cuda_obj)}
+            await asyncio.sleep(1)
+            frames = await to_frames(msg, serializers=("cuda", "dask", "pickle"))
+
+            # Send meta data
+            await ep.send(np.array([len(frames)], dtype=np.uint64))
+            await ep.send(
+                np.array(
+                    [hasattr(f, "__cuda_array_interface__") for f in frames],
+                    dtype=np.bool,
+                )
+            )
+            await ep.send(np.array([nbytes(f) for f in frames], dtype=np.uint64))
+            # Send frames
+            for frame in frames:
+                if nbytes(frame) > 0:
+                    await ep.send(frame)
+
+            print("CONFIRM RECEIPT")
+            close_msg = b"shutdown listener"
+            msg_size = np.empty(1, dtype=np.uint64)
+            await ep.recv(msg_size)
+
+            msg = np.empty(msg_size[0], dtype=np.uint8)
+            await ep.recv(msg)
+            recv_msg = msg.tobytes()
+            assert recv_msg == close_msg
+            print("Shutting Down Server...")
+            await ep.signal_shutdown()
+            ep.close()
+            lf.close()
+
+        lf = ucp.create_listener(write, port=listener_port)
+        try:
+            while not lf.closed:
```

Comment:
@madsbk this while loop seems wrong.  in the old bindings  lf had a coroutine accessor attached to it and I could hand that back to the loop.  should I be doing something differently here ?

Reply:
Maybe save `lf` globally then it will not go out of scope?